osquery/tables/networking/windows/windows_firewall_rules.cpp

```diff
@@ -0,0 +1,411 @@
+         "Failed to get firewall rule local ports"},
+        {WindowsFirewallError::RuleRemotePortsError,
+         "Failed to get firewall rule remote ports"},
+        {WindowsFirewallError::RuleICMPTypesCodesError,
+         "Failed to get firewall rule ICMP types and codes"},
+        {WindowsFirewallError::RuleProfilesError,
+         "Failed to get firewall rule profiles"},
+    };
+
+std::string getErrorDescription(const WindowsFirewallError& error) {
+  auto it = kWindowsFirewallErrorDescriptions.find(error);
+  if (it == kWindowsFirewallErrorDescriptions.end()) {
+    std::stringstream stream;
+    stream << "Unknown error type: 0x" << std::hex
+           << static_cast<std::uint64_t>(error);
+
+    return stream.str();
+  }
+
+  return it->second;
+}
+
+Expected<WindowsFirewallRule, WindowsFirewallError>
+createWindowsFirewallRuleError(WindowsFirewallError error, HRESULT hr) {
+  return createError(error)
+         << getErrorDescription(error) << ", HRESULT=0x" << std::hex << hr;
+}
+
+Expected<WindowsFirewallRules, WindowsFirewallError>
+createWindowsFirewallRulesError(WindowsFirewallError error, HRESULT hr) {
+  return createError(error)
+         << getErrorDescription(error) << ", HRESULT=0x" << std::hex << hr;
+}
+
+Row renderFirewallRule(const WindowsFirewallRule& rule) {
+  Row r;
+
+  r["name"] = rule.name;
+  r["app_name"] = rule.appName;
+
+  switch (rule.action) {
+  case NET_FW_ACTION_BLOCK:
+    r["action"] = "Block";
+    break;
+  case NET_FW_ACTION_ALLOW:
+    r["action"] = "Allow";
+    break;
+  default:
+    r["action"] = "";
+    break;
+  }
+
+  r["enabled"] = INTEGER(rule.enabled);
+
+  switch (rule.direction) {
+  case NET_FW_RULE_DIR_IN:
+    r["direction"] = "In";
+    break;
+  case NET_FW_RULE_DIR_OUT:
+    r["direction"] = "Out";
+    break;
+  default:
+    r["direction"] = "";
+    break;
+  }
+
+  switch (rule.protocol) {
+  case NET_FW_IP_PROTOCOL_TCP:
+    r["protocol"] = "TCP";
+    break;
+  case NET_FW_IP_PROTOCOL_UDP:
+    r["protocol"] = "UDP";
+    break;
+  case NET_FW_IP_PROTOCOL_ANY:
+    r["protocol"] = "Any";
+    break;
+  default:
+    r["protocol"] = "";
+    break;
+  }
+
+  r["local_addresses"] = rule.localAddresses;
+  r["remote_addresses"] = rule.remoteAddresses;
+
+  if (rule.protocol != NET_FW_IP_VERSION_V4 &&
+      rule.protocol != NET_FW_IP_VERSION_V6) {
+    r["local_ports"] = rule.localPorts;
+    r["remote_ports"] = rule.remotePorts;
+    r["icmp_types_codes"] = "";
+  } else {
+    r["local_ports"] = "";
+    r["remote_ports"] = "";
+    r["icmp_types_codes"] = rule.icmpTypesCodes;
+  }
+
+  r["profile_domain"] =
+      INTEGER((rule.profileBitmask & NET_FW_PROFILE2_DOMAIN) != 0);
+  r["profile_private"] =
+      INTEGER((rule.profileBitmask & NET_FW_PROFILE2_PRIVATE) != 0);
+  r["profile_public"] =
+      INTEGER((rule.profileBitmask & NET_FW_PROFILE2_PUBLIC) != 0);
+
+  return r;
+}
+
+typedef HRESULT (STDMETHODCALLTYPE INetFwRule::*BSTRFunc)(BSTR*);
```

Comment:
`using` should be used instead of `typedef`: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rt-using